src/kernel-install/kernel-install.c

```diff
@@ -0,0 +1,1166 @@
+
+        r = getenv_bool("KERNEL_INSTALL_BYPASS");
+        if (r <= 0) {
+                if (r != -ENXIO)
+                        log_debug_errno(r, "Failed to parse $KERNEL_INSTALL_BYPASS, assuming no.");
+                return false;
+        }
+
+        log_debug("$KERNEL_INSTALL_BYPASS is enabled, skipping execution.");
+        return true;
+}
+
+static int help(void) {
+        _cleanup_free_ char *link = NULL;
+        int r;
+
+        r = terminal_urlify_man("kernel-install", "8", &link);
+        if (r < 0)
+                return log_oom();
+
+        printf("%1$s [OPTIONS...] COMMAND ...\n\n"
+               "%2$sAdd and remove kernel and initrd images to and from /boot%3$s\n"
+               "\nUsage:\n"
+               "  %1$s [OPTIONS...] add KERNEL-VERSION KERNEL-IMAGE [INITRD-FILE...]\n"
+               "  %1$s [OPTIONS...] remove KERNEL-VERSION\n"
+               "  %1$s [OPTIONS...] inspect\n"
+               "\nOptions:\n"
+               "  -h --help              Show this help\n"
+               "     --version           Show package version\n"
+               "  -v --verbose           Increase verbosity\n"
+               "\nSee the %4$s for details.\n",
+               program_invocation_short_name,
+               ansi_highlight(),
+               ansi_normal(),
+               link);
+
+        return 0;
+}
+
+static int parse_argv(int argc, char *argv[]) {
+        enum {
+                ARG_VERSION = 0x100,
+        };
+        static const struct option options[] = {
+                { "help",                 no_argument,       NULL, 'h'                      },
+                { "version",              no_argument,       NULL, ARG_VERSION              },
+                { "verbose",              no_argument,       NULL, 'v'                      },
+                {}
+        };
+        int t;
+
+        assert(argc >= 0);
+        assert(argv);
+
+        while ((t = getopt_long(argc, argv, "hv", options, NULL)) >= 0)
+                switch (t) {
+                case 'h':
+                        return help();
+
+                case ARG_VERSION:
+                        return version();
+
+                case 'v':
+                        log_set_max_level(LOG_DEBUG);
+                        arg_verbose = true;
+                        break;
+
+                case '?':
+                        return -EINVAL;
+
+                default:
+                        assert_not_reached();
+                }
+
+        return 1;
+}
+
+static int run(int argc, char* argv[]) {
+        static const Verb verbs[] = {
+                { "add",         3,        VERB_ANY, 0,            verb_add            },
+                { "remove",      2,        2,        0,            verb_remove         },
+                { "inspect",     1,        1,        VERB_DEFAULT, verb_inspect        },
+                {}
+        };
+        _cleanup_(context_done) Context c = {
+                .rfd = AT_FDCWD,
```

Comment:
Let's do the same as firstboot and always open a file descriptor to the root directory so that we don't run into trouble if we accidentally use a relative path somewhere with chaseat().

Reply:
I intended to set AT_FDCWD here, as `$KERNEL_INSTALL_PLUGINS` supported relative path. But, the env var is for debugging and testing, so we can change the behavior. I will change as suggested.

Reply:
One more big reason I forgot is, if we set FD of the root directory, then the result of `chaseat()` is always relative, and we need to add prefix "/" even --root= is not specified. That's super inconvenient. Maybe, we should add `CHASE_ABSOLUTE` or so?

Reply:
Yeah I thought about the same thing, `CHASE_ABSOLUTE` makes sense. We already have `need_absolute` in `chaseat()`, should be trivial to plug in there.

We should probably refuse the flag in all of the chaseat() helpers, though.

Reply:
Actually, I'm not sure if that's the way to go after all.

What about making chaseat() return absolute paths if `dir_fd_is_root_or_cwd()` instead of checking for `AT_FDCWD` instead?

Reply:
→ #27299.